avocado/core/requirements.py

```diff
@@ -0,0 +1,94 @@
+
+import argparse
+import os
+
+from . import data_dir
+from . import safeloader
+from ..utils.asset import Asset
+from ..utils.vmimage import get as get_vmimage
+from ..utils import data_structures
+
+
+class RequirementsResolver:
+
+    def __init__(self, references):
```

Comment:
My initial reaction is that this resolver should act on a `avocado.core.nrunner.Task`.  Right now we have `check_task_requirements()` ( https://github.com/avocado-framework/avocado/blob/master/avocado/core/nrunner.py#L38 ) which could be replaced or augmented by this RequirementResolver.

Then, the RequirementResolver could be part of the scheduling/spawning of tasks.  The following pseudocode should help to illustrate it:

```python

 j = Job(references)
 # from this point on, j.test_suite contains a list of Tasks
```

The `run()` implementation of a `Job` could look like:

```python

def run(self):
   requirements_resolver = RequirementsResolver()
   while True:
      try:
          task = self.test_suite.pop()
      except ValueError:
         break
      if requirements_resolver.check_task_requirements(task):
         self.spawner.spawn(task)
      else:
         requirements_resolver.resolve_task_requirement_in_the_background(task)
```